sdks/java/extensions/sql/src/test/java/org/apache/beam/sdk/extensions/sql/meta/provider/kafka/BeamKafkaTableTest.java

```diff
@@ -0,0 +1,170 @@
+
+  /** Returns proper implementation of BeamKafkaTable for the tested format */
+  protected abstract BeamKafkaTable getBeamKafkaTable();
+
+  /** Returns encoded payload for the tested format. */
+  protected abstract byte[] generateEncodedPayload(int i);
+
+  /** Provides a deterministic row from the given integer. */
+  protected abstract Row generateRow(int i);
+
+  @Test
+  public void testOrderedArrivalSinglePartitionRate() {
+    KafkaTestTable table = getTestTable(1);
+    for (int i = 0; i < 100; i++) {
+      table.addRecord(createKafkaTestRecord("k" + i, i, 500L * i));
+    }
+
+    BeamTableStatistics stats = table.getTableStatistics(null);
+    Assert.assertEquals(2d, stats.getRate(), 0.001);
+  }
+
+  @Test
+  public void testOrderedArrivalMultiplePartitionsRate() {
+    KafkaTestTable table = getTestTable(3);
+    for (int i = 0; i < 100; i++) {
+      table.addRecord(createKafkaTestRecord("k" + i, i, 500L * i));
+    }
+
+    BeamTableStatistics stats = table.getTableStatistics(null);
+    Assert.assertEquals(2d, stats.getRate(), 0.001);
+  }
+
+  @Test
+  public void testOnePartitionAheadRate() {
+    KafkaTestTable table = getTestTable(3);
+    for (int i = 0; i < 100; i++) {
+      table.addRecord(createKafkaTestRecord("1", i, 1000L * i));
+      table.addRecord(createKafkaTestRecord("2", i, 500L * i));
+    }
+
+    table.setNumberOfRecordsForRate(20);
+    BeamTableStatistics stats = table.getTableStatistics(null);
+    Assert.assertEquals(1d, stats.getRate(), 0.001);
+  }
+
+  @Test
+  public void testLateRecords() {
+    KafkaTestTable table = getTestTable(3);
+
+    table.addRecord(createKafkaTestRecord("1", 132, 1000L));
+    for (int i = 0; i < 98; i++) {
+      table.addRecord(createKafkaTestRecord("1", i, 500L));
+    }
+    table.addRecord(createKafkaTestRecord("1", 133, 2000L));
+
+    table.setNumberOfRecordsForRate(200);
+    BeamTableStatistics stats = table.getTableStatistics(null);
+    Assert.assertEquals(1d, stats.getRate(), 0.001);
+  }
+
+  @Test
+  public void testAllLate() {
+    KafkaTestTable table = getTestTable(3);
+
+    table.addRecord(createKafkaTestRecord("1", 132, 1000L));
+    for (int i = 0; i < 98; i++) {
+      table.addRecord(createKafkaTestRecord("1", i, 500L));
+    }
+
+    table.setNumberOfRecordsForRate(200);
+    BeamTableStatistics stats = table.getTableStatistics(null);
+    Assert.assertTrue(stats.isUnknown());
+  }
+
+  @Test
+  public void testEmptyPartitionsRate() {
+    KafkaTestTable table = getTestTable(3);
+    BeamTableStatistics stats = table.getTableStatistics(null);
+    Assert.assertTrue(stats.isUnknown());
+  }
+
+  @Test
+  public void allTheRecordsSameTimeRate() {
+    KafkaTestTable table = getTestTable(3);
+    for (int i = 0; i < 100; i++) {
+      table.addRecord(createKafkaTestRecord("key" + i, i, 1000L));
+    }
+    BeamTableStatistics stats = table.getTableStatistics(null);
+    Assert.assertTrue(stats.isUnknown());
+  }
```

Comment:
The tests above this point are the only ones that use `KafkaTestTable`, and they're only exercising the `getTableStatistics` method, which never deserializes any records. So:

1. We shouldn't really need to repeat these tests for each payload format.
2. We don't need a separate `KafkaTestTableCSV` and `KafkaTestTableAvro`. There could just be a single concrete `KafkaTestTable` that raises an error in `getPTransformFor{Input,Output}`.

I'm fine if we don't worry about (1) for now, but I'd like to address (2) for clarity.